src/java/org/apache/cassandra/index/sai/disk/v1/PartitionAwarePrimaryKeyFactory.java

```diff
@@ -72,10 +65,12 @@ private PartitionAwarePrimaryKey(Token token, DecoratedKey partitionKey, Supplie
         @Override
         public PrimaryKey loadDeferred()
         {
-            if (primaryKeySupplier != null && partitionKey == null)
+            if (primaryKeySupplier != null)
             {
+                assert partitionKey == null : "By definition cannot be otherwise";
                 this.partitionKey = primaryKeySupplier.get().partitionKey();
                 primaryKeySupplier = null;
```

Comment:
Side note: 
Just realized this class is not thread safe because of this.
I hope we're not using this in multithreaded context. 
I don't recall us doing that fortunately.

Can you add `@NotThreadSafe` on top of it?